components/plan-build/bin/hab-plan-build.sh

```diff
@@ -870,6 +870,22 @@ _return_or_append_to_set() {
   return 0
 }
 
+# **Internal** Appends an entry to the given array and removes any entries
+# already in the array that match the passed entry. Note that this is specifically
+# intended for lists of dependencies used for building TDEPS metadata. This should
+# not be used as a generic array appender because it could have unexpected results
+# with arbitrary data.
+_add_dep_to_tdep_list() {
+  local appended_set
+  local dedupe_set
+  local deps
+  deps=("${@:2}")
+  dedupe_set=("${deps[@]/$1}")
+  appended_set=("${dedupe_set[@]}" "$1")
+  echo "${appended_set[@]}"
+  return 0
+}
```

Comment:
In light of the subtleties that @stevendanna raised, it may be better to be more explicit here:

```bash
_add_dep_to_tdep_list() {
  local to_add="${1}"
  local deps=("${@:2}")
  local result=()

  # Explicitly filter out any instances of the to-be-added dependency
  # that we may have already seen.
  for d in "${deps[@]}"; do
      if [[ "$d" != "${to_add}" ]]; then
          result=( "${result[@]}" "$d" )
      fi
  done
  # Append the dependency to the end of the list
  result=( "${result[@]}" "$to_add" )

  echo "${result[@]}"
  return 0
}
```
It's wordier, and perhaps less efficient (though I'd be surprised to discover this being a bottleneck in the overall process of building a Habitat package). However, it's _much_ more obvious what is going on (I _always_ have to look up these variable manipulation Bash-isms), and isn't subject to the partial replacement issues the other implementation is. Additionally, if we ever get around to rewriting this in Rust, it will be much easier to translate this explicit implementation than the subtle "idiomatic" other one.

Reply:
Incidentally, as a point of Bash style, I personally find it much easier to reason about functions when the arguments are immediately bound to descriptive variable names; `$to_add` is better than `$1`.

Reply:
I like it

Reply:
I have made this change @christophermaier and rebuilt everything and reran locally against the e2e tests which passed.